src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs

```diff
@@ -33368,635 +33971,3631 @@ .maxstack  3
+                Diagnostic(ErrorCode.ERR_BadArgType, "..y").WithArguments("1", "int", "R").WithLocation(21, 36),
+                // (21,38): error CS1950: The best overloaded Add method 'MyCollection<int?>.Add(R)' for the collection initializer has some invalid arguments
+                //         MyCollection<int?> z = [x, ..y, null];
+                Diagnostic(ErrorCode.ERR_BadArgTypesForCollectionAdd, "y").WithArguments("MyCollection<int?>.Add(R)").WithLocation(21, 38),
+                // (21,41): error CS1950: The best overloaded Add method 'MyCollection<int?>.Add(R)' for the collection initializer has some invalid arguments
+                //         MyCollection<int?> z = [x, ..y, null];
+                Diagnostic(ErrorCode.ERR_BadArgTypesForCollectionAdd, "null").WithArguments("MyCollection<int?>.Add(R)").WithLocation(21, 41),
+                // (21,41): error CS1503: Argument 1: cannot convert from '<null>' to 'R'
+                //         MyCollection<int?> z = [x, ..y, null];
+                Diagnostic(ErrorCode.ERR_BadArgType, "null").WithArguments("1", "<null>", "R").WithLocation(21, 41));
+        }
+
+        [Theory]
+        [InlineData("out")]
+        [InlineData("ref")]
+        public void AddMethod_ByRef_Extension_05(string refKind)
+        {
+            string source = $$"""
+                using N;
+                using System.Collections;
+                using System.Collections.Generic;
+                ref struct R { }
+                class MyCollection<T> : IEnumerable
+                {
+                    private readonly List<T> _list = new();
+                    IEnumerator IEnumerable.GetEnumerator() => _list.GetEnumerator();
+                    internal void __AddInternal(T t) { _list.Add(t); }
+                    public void Add(R r) => throw null;
+                }
+                static class Extensions
+                {
+                    public static void Add<T>(this MyCollection<T> collection, {{refKind}} T t) { t = default; collection.__AddInternal(t); }
+                }
+                namespace N
+                {
+                    static class Extensions
+                    {
+                        public static void Add<T>(this MyCollection<T> collection, T t) { collection.__AddInternal(t); }
+                    }
+                }
+                class Program
+                {
+                    static void Main()
+                    {
+                        int x = 1;
+                        int[] y = [2, 3];
+                        MyCollection<int?> z = [x, ..y, null];
+                        z.Report();
+                    }
+                }
+                """;
+            CompileAndVerify([source, s_collectionExtensions], expectedOutput: "[1, 2, 3, null], ");
+        }
+
+        [Theory]
+        [InlineData("")]
+        [InlineData("in")]
+        [InlineData("ref readonly")]
```

Comment:
Why does this work for `ref readonly`? I would, at the very least, expect a warning?

Reply:
Same response, I guess.